File: src/server/src/test/test_velvet_dawn/test_mechanics/test_actions/test_get_action.py
import velvet_dawn.datapacks
from velvet_dawn import errors
from velvet_dawn.dao import app
from test.base_test import BaseTest
from velvet_dawn.mechanics.actions import ActionModify
from velvet_dawn.models.coordinate import Coordinate


class TestGetAction(BaseTest):

    def test_get_actions(self):
        action = velvet_dawn.mechanics.actions.get_action("0", {
            "modify": "self.health",
            "set": "x"
        })
        self.assertTrue(isinstance(action, ActionModify))

        # Invalid action
        with self.assertRaises(errors.ValidationError):
            velvet_dawn.mechanics.actions.get_action("0", {
                "random key": "self.health",
            })

    def test_action_can_run(self):
        with app.app_context():
            self.prepare_game()

            unit = velvet_dawn.db.units.get_units_at_positions(Coordinate(5, 0))[0]
            action = velvet_dawn.mechanics.actions.get_action("0", {
                "modify": "self.health",
                "set": 100,
                "conditions": [{
                    "if": "self.health",
                    "lte": 10
                }]
            })

            unit.set_attribute("health", 50)
            self.assertFalse(action.can_run(unit))

            unit.set_attribute("health", 2)
            self.assertTrue(action.can_run(unit))

File: src/jserver/src/test/velvetdawn/mechanics/actions/TestGetAction.java
package test.velvetdawn.mechanics.actions;

import org.junit.Test;
import test.BaseTest;
import velvetdawn.mechanics.actions.ActionModify;
import velvetdawn.mechanics.actions.Actions;
import velvetdawn.models.Coordinate;
import velvetdawn.models.anytype.Any;
import velvetdawn.models.anytype.AnyJson;
import velvetdawn.models.anytype.AnyList;
import velvetdawn.models.instances.entities.EntityInstance;

import java.util.List;

import static junit.framework.TestCase.*;
import static org.junit.Assert.assertThrows;


public class TestGetAction extends BaseTest {

    @Test
    public void testGetActions() throws Exception {
        var velvetDawn = this.prepareGame();

        var action = Actions.fromJson(velvetDawn, "0", new AnyJson()
                .set("modify", "self.health")
                .set("set", "x"));

        assertTrue(action instanceof ActionModify);

        // Invalid action
        assertThrows(Exception.class, () -> {
            Actions.fromJson(velvetDawn, "0", new AnyJson().set("random key", "self.health"));
        });
    }

    @Test
    public void testActioCanRun() throws Exception {
        var velvetDawn = this.prepareGame();

        EntityInstance unit = velvetDawn.entities.getAtPosition(new Coordinate(5, 0)).get(0);
        var action = Actions.fromJson(velvetDawn, "0", new AnyJson()
                .set("modify", "self.health")
                .set("set", 100)
                .set("conditions", AnyList.of(
                        new AnyJson()
                                .set("if", "self.health")
                                .set("lte", 10)
                )));

        unit.attributes.set("health", Any.from(50));
        assertFalse(action.canRun(unit));

        unit.attributes.set("health", Any.from(2));
        assertTrue(action.canRun(unit));
    }
}

